부산대 BE_이시웅 3단계 - 데이터베이스 적용

silverttthin opened the pull request:
크게 3가지 구현사항이 있습니다.
1. 기존 hashmap 기반 db 로직을 jdbctemplate 기반으로 전면 교체했습니다.
    - preparedStatement 등을 다루면서 과거의 서버 개발에선 sql을 잘 알고 코드를 잘 짜야했다는 것을 좀 느낀 것 같습니다. 
    - 저처럼 먼저, 그리고 오직 jpa만 배운  ~~MZ한~~ 스프링부트 학습자들은 꽤 불편함을 느낀 경험인 것 같습니다. JPA가 확실히 백엔드 개발에서 sql을 다루는 비중을 확실히 줄였다는게 이번 step3 역사공부를 통해 체감된 것 같습니다.

2. validation 종속성을 설치해서 ItemRequest dto에 간단한 검증로직을 어노테이션으로 넣었습니다.
    - 학습과정 맞춰 나중에 도입할까 했었는데.. 써본 경험이 있어서 단순한 if문보다 훨씬 편한걸 느낀 경험이 있어서 그냥 일찍 도입했습니다.
    - rest controller와 admin controller 둘다 validation 적용되게끔 컨트롤러 인자에 `@valid`를 명시해뒀습니다.

3. PostConsturct로 서비스 빈 초기화시 수행하던 DB drop/create 및 목업 데이터 삽입 로직을 삭제하고 schema.sql, data.sql을 작성했습니다.


+) pr과 관계없이 리뷰어님께 드리는 개인적인 질문
이번 여수 카카오 커넥트 데이에서 카카오 개발자분께 제가 다음 질문을 드리고 답변받았습니다.
Q: 레이어드 아키텍처로 적당히 api 서버를 만들고 배포를 해 프로젝트를 해서 포트폴리오에 내세우는건 llm이 보편화되고 채용이 심화된 2025년엔 개인적으로 전국구 단위 경쟁시장에서 많이 부족한 수준일 것 같다. 좀더 심화 학습을 해야할 것 같은 필요성을 개인적으로 크게 느낀다. 채용자 입장에서 신입 지원 학부생을 기술적 관점으로 평가할 때 수직적 확장(서비스 로직에 스레드 도입, 쿼리 ms 단축), 수평적 확장(개인 스택에 백엔드 말고도 AI, 클라우드 같은 다양한 기술스택 부각) 중 뭐가 나을까요? 

A: 무엇을 했냐보다 중요한건 이 문제를 왜? 어떻게? 해결하고 그 고민한 과정, 팀원과 논의한 스토리를 더 듣고싶다. 어차피 합격해도 6개월 수습기간 거친다. 그 기간동안 잘 배우고 잘 할 사람을 뽑는다. 즉 문제를 어떻게 풀어나갔는지를 잘 보여주고 잘하는 사람을 더 좋아할 것 같다

결국 ai가 나와도 본질은 문제를 잘 정의하고 기술적으로 이 문제를 왜? 어떻게? 푸는 것이라는 건 변함이 없다는 깨달음을 얻고 온 귀중한 시간이었던 것 같습니다. 혹시 리뷰어님의 개인적인 의견도 여쭤봐도 될까요? 추가로! 그럼에도 수직적 확장 방법에 조금 관심, 궁금증이 있긴 합니다! 디자인 패턴이라거나 스레드 적용 등 이 부분에 대해 현업에서 어떤 기능에서 어떤 방법을 왜 적용했는지 매우 궁금합니다

silverttthin committed:
chore: dto 폴더로 분류

silverttthin committed:
docs: 디비 마이그레이션 해야할 것들 추가

silverttthin committed:
feat: jdbcTemplate 연결 및 목업 데이터 15개 추가를 jdbc update 메서드로 변경

silverttthin committed:
docs: 요구사항 세분화

silverttthin committed:
feat[jdbc]: 게시글 전체/단건 조회
-  지피티를 지양하고 최대한 공식문서와 블로그를 참고하며 구현해보았다.
- query(쿼리, rowmapper)로 전체조회 <- preparedStatement 아님
- queryForObject(쿼리, rowmapper, {변수})로 단건 조회 <- preparedStatement!

silverttthin committed:
docs: 검증로직 todo에 추가

silverttthin committed:
feat[jdbc]: 생성 기능 구현
- 기존 createItem 서비스 함수는 generated pk를 리턴
- jdbc에서 이를 리턴하기위해 다음 2가지를 수행
1. preparedStatement 객체에 RETURN_GENERATED_KEYS 명시
2. 키 값을 받아올 keyHolder를 명시
(꽤 번거로운 작업인 것 같다. jpa는 save함수 리턴을 이용하면 매우 쉬웠는데)

silverttthin committed:
feat[jdbc]: 수정/삭제기능 구현
- 단순히 jdbctemplate.update는 preparedStatement와 ?인자에 들어갈 값들을 넣어주면 된다.
- jdbctemplate는 쿼리 스트링을 만들고 실행시키는데에는 편리하나 뭐랄까 쿼리 실행이후 무언가를 리턴하는 작업을 하기엔 좀 번거로운 면이 있다.

silverttthin committed:
build: validation 종속성 설치

silverttthin committed:
feat: ItemRequestDto에 입력값 검증 추가

silverttthin committed:
feat: ItemRequestDto에 입력값 검증 추가

silverttthin committed:
remove: db 초기화 및 목업데이터 삽입 코드 삭제

silverttthin committed:
refactor: db 초기화 및 목업 데이터 삽입 로직 변경
- 해결하려는 문제를 왜? 어떻게로 분석하는 습관을 가져보는 차원에서
왜?: 기존 서비스레이어에서 디비를 초기화하는건 일단 자체부터가 부자연스럽고 SRP도 위배
어떻게?: ddl과 초기화 로직을 각각 schema/data.sql에 명시하고 resources에 저장해두기 -> 스프링이 알아서 실행시켜줌

silverttthin committed:
remove: item entity에서 auto increment 수작업 기능 삭제

silverttthin committed:
feat: dto 상품명 검증로직 세분화
- 카카오 <- 이건 다음 step에서 구현할 예정

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -14,59 +20,66 @@
-			db.put(item.getId(), item);
-		}
-	}
+	private final JdbcTemplate jdbcTemplate;
```
jdbctemplate 을 사용해주셨군요 👍 

> 저처럼 먼저, 그리고 오직 jpa만 배운 MZ한 스프링부트 학습자들은 꽤 불편함을 느낀 경험인 것 같습니다. JPA가 확실히 백엔드 개발에서 sql을 다루는 비중을 확실히 줄였다는게 이번 step3 역사공부를 통해 체감된 것 같습니다.

역사공부.. 🤣🤣
JPA도 좋지만, 진짜 실무에서는 SQL 먼저 익히는 게 훨씬 도움돼요!
나중에 JPA 쓸 때도 SQL 원리 알아야 트러블슈팅 할 수 있고,
슬로우 쿼리나 인덱스 같은 거 볼 때도 SQL 모르면 막히거든요.
그래서 순서는 SQL → JPA 추천합니다! 👍

mskangg commented on src/main/java/gift/product/ItemController.java:
```diff
@@ -18,7 +21,7 @@ public class ItemController {
 	// 게시글 생성
 	@PostMapping()
-	public Long createItem(@RequestBody ItemRequest req) {
+	public Long createItem(@Valid @RequestBody ItemRequest req) {
```
👍

mskangg commented on README.md:
```diff
@@ -24,4 +24,14 @@
\ No newline at end of file
+- [X] 상품 삭제 기능 연동
+
+### [✔️] DB 마이그레이션 (HashMap -> h2)
+- [X] h2 연결 및 목업 데이터 추가
+- [X] jdbctemplate 메서드로 기존 코드 리팩토링
+    - [X] 게시글 조회 
+    - [X] 게시글 생성
+    - [X] 게시글 수정
+    - [X] 게시글 삭제
+- [X] validation 추가 
+- [X] db 초기화 및 목업데이터 삽입 로직 위치 변경 (service -> schema/data.sql) 
```
요구사항 정리 👍

mskangg commented on src/main/java/gift/product/dto/ItemRequest.java:
```diff
@@ -0,0 +1,22 @@
+
+	@NotBlank
+	@Size(max = 15)
+	@Pattern(
+		regexp = "^[A-Za-z가-힣0-9()\\[\\]\\+\\-\\&/_]+$",
+		message = "영문·숫자 및 한글, 그리고 다음 특수기호 ()[]+-&/_ 만 허용합니다."
+	)
```
정규식까지 👍

mskangg commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,16 @@
+INSERT INTO item (name, price, image_url) VALUES
+('item1', 1000, 'url1'),
+('item2', 2000, 'url2'),
+('item3', 3000, 'url3'),
+('item4', 4000, 'url4'),
+('item5', 5000, 'url5'),
+('item6', 6000, 'url6'),
+('item7', 7000, 'url7'),
+('item8', 8000, 'url8'),
+('item9', 9000, 'url9'),
+('item10', 10000, 'url10'),
+('item11', 11000, 'url11'),
+('item12', 12000, 'url12'),
+('item13', 13000, 'url13'),
+('item14', 14000, 'url14'),
+('item15', 15000, 'url15');
```
👍

mskangg commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,8 @@
+DROP TABLE IF EXISTS item;
+
+CREATE TABLE item (
+                      id BIGINT PRIMARY KEY AUTO_INCREMENT,
+                      name VARCHAR(255) NOT NULL,
+                      price INT NOT NULL,
+                      image_url VARCHAR(1000) NOT NULL
+);
```
👍

mskangg commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,16 @@
+('item12', 12000, 'url12'),
+('item13', 13000, 'url13'),
+('item14', 14000, 'url14'),
+('item15', 15000, 'url15');
```
미션의 내용과는 별개지만 참고하시라고 코멘트 남깁니다.

<img width="270" alt="image" src="https://github.com/user-attachments/assets/7f76cab5-175e-40a2-aa4c-7b93cb44356b" />

newline을 넣어주시면 좋을 것 같아요!

Github에서는 newline을 남기지 않으면 위와 같이 표식이 남습니다.

POSIX 표준에 맞지 않기도 하고 작업시 newline을 붙이는데, 다른 작업자가 붙이지 않을 경우 불필요한 file change가 생기기도 합니다.

저 같은 경우 IntelliJ에서 newline을 자동으로 추가하는 설정을 사용하고 있습니다.

<img width="702" alt="image" src="https://github.com/user-attachments/assets/989b6519-35fb-4d8b-a63c-d5b83de5bdfb" />

`Ensure every saved file ends with a line break`

silverttthin replied:
감사합니다 보자마자 설정켜서 적용했습니다

mskangg commented on src/main/resources/application.properties:
```diff
@@ -1,2 +1,6 @@
+# h2-console ??? ??
+spring.h2.console.enabled=true
+# db url
+spring.datasource.url=jdbc:h2:mem:test
```
👍

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -14,59 +20,66 @@
-	}
+	private final JdbcTemplate jdbcTemplate;
+
+	public ItemService(JdbcTemplate jdbcTemplate) {this.jdbcTemplate = jdbcTemplate;}
```
Service 에서 데이터베이스 영속성을 갖게 하셨군요 🤔
Repository 계층을 두고 분리하는게 좋을것 같은데, 어떤점이 좋을까요?

silverttthin replied:
이전 리뷰때 말씀드린것처럼 클래스의 책임을 덜고 개별 집중하는데에 이점을 얻을 것 같습니다.
- 서비스 -> 서비스 로직, 검증로직
- 리포지토리 -> 데이터베이스 관련 로직

기존 로직을 jdbc로 리팩토링할땐 그것에 집중해 느끼지 못했는데 지금보니 서비스 레이어에서 디비로직이 꽤 많은 부분을 차지하고 있어 슬 분할할 필요성이 느껴지는 것 같습니다

silverttthin replied:
한가지 궁금한게 있는데 dao와 repository가 결국 디비 로직에 대한 책임을 갖는 클래스라는 점에서 차이를 느끼기가 어려웠습니다. 굳이 따지자면 후자는 전자보다는 레이어로서 더 아키테쳐틱한 개념을 갖는 것 같다는 느낌..?
따로 검색해서 찾아봤는데 인프런 답변, 스택오버플로우 등에 비슷한 질문들이 있었고 대체로 실제 개발에선 별 차이없이 거의 같다고 봐도 무방하다고 하였는데 맞을까요?

mskangg replied:
@silverttthin 

DAO와 Repository는 실무에서 두 용어가 거의 비슷하게 쓰이긴 하지만, 약간의 개념적 차이가 있어요 🤔
DAO는 주로 데이터베이스 테이블 중심으로, CRUD 같은 세밀한 작업을 담당하는 저수준 계층이에요 (UserDao가 user 테이블만 직접 다룸)
Repository는 도메인 객체(엔티티) 중심으로, 객체의 집합을 관리하는 고수준 계층입니다 🤔

실제로는 두 패턴이 거의 비슷하게 사용되고, JPA 환경에서는 Repository라는 이름이 표준처럼 자리잡았어요
그래서 `실질적 차이는 거의 없다`는 의견도 많지만, 지금은 `DAO는 DB 테이블, 저수준, Repository는 도메인, 고수준`이라는 정도로 구분해두면 좋을 것 같아요 👍 

결론적으로, 작은 프로젝트나 단순 CRUD 위주라면 DAO, Repository 구분 없이 써도 무방하고
도메인 중심 설계나 복잡한 비즈니스가 있다면 Repository 패턴을 더 추천합니다!

추가로, Spring의 `@Repository` 어노테이션 주석에도 아래와 같이 설명이 되어 있어요
[Spring @Repository Javadoc](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/stereotype/Repository.html)

> Indicates that an annotated class is a "Repository", originally defined by Domain-Driven Design (Evans, 2003) as a mechanism for encapsulating storage, retrieval, and search behavior which emulates a collection of objects.
Teams implementing traditional Jakarta EE patterns such as "Data Access Object" may also apply this stereotype to DAO classes, though care should be taken to understand the distinction between Data Access Object and DDD-style repositories before doing so. This annotation is a general-purpose stereotype and individual teams may narrow their semantics and use as appropriate.

즉, `@Repository`는 DDD에서 말하는 `저장소` 개념을 표현하고, DAO 패턴에도 쓸 수 있지만, DDD style Repository와는 역할에 약간 차이가 있을 수 있다고 명시하고 있어요.
그리고 `@Repository`로 마킹된 클래스는 Spring의 예외 변환(DataAccessException translation) 같은 부가 기능도 자동으로 적용됩니다 👍

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -14,59 +20,66 @@
-		item.setPrice(req.price());
-		item.setImageUrl(req.imageUrl());
-		return new GetItemResponse(item.getId(), item.getName(), item.getPrice(), item.getImageUrl());
+		final String sql = "update item set name = ?, price = ?, image_url = ? where id = ?";
```
상품을 수정할 때 쿼리를 바로 호출하기보단,
먼저 상품을 조회해서 수정에 필요한 비즈니스 로직(유효성 검사, 상태 체크 등)을 거친 뒤
다시 저장하는 방식이 더 좋습니다 👍
이렇게 되면, 비즈니스 규칙을 제대로 적용할 수 있고 객체지향적으로 도메인 로직을 활용할 수 있어서 유지보수나 확장에도 유리하거든요 😊

데이터베이스 계층에선 CRUD 를 행하는 로직만 작성하고 비즈니스는 분리하는건 어떨까요?
```java
public void updateItem(Item item) {
}
```

silverttthin replied:
이렇게 되면 궁금해지는게 생겼습니다
상품 수정 서비스 함수는 일단 그 입력으로 들어온 상품 id가 실제 존재하는지 확인위해 조회를 하는 절차는 필수일텐데 이러면 수정은 조회 함수에 의존하게 될 것 같은데..

사실 당연히 필요한 로직이긴한데 책임의 분리가 제대로 이루어지는건 아닐 것 같아 쪼금 걸리는 부분입니다. 큰 문제가 없는 부분일까요?

mskangg replied:
@silverttthin 

Repository는 데이터 접근만 담당하는 계층이라, 책임이 분리되는 게 맞아요.
서비스에서 조회와 수정 로직이 같이 있는 게 혼동될 수 있지만, 실제로는 서비스가 비즈니스 요구사항(존재 확인, 유효성 검사, 상태 변경 등)을 하나의 흐름으로 처리하는 게 자연스러운 구조입니다 👍 

즉, Repository는 데이터 CRUD만 Service는 비즈니스 로직(여러 데이터 접근, 도메인 규칙 적용 등) 이렇게 역할이 명확히 나뉘어 있는 것이 책임분리에요 🤔

서비스에서 `조회 → 비즈니스 처리 → 저장`이 한 트랜잭션/로직 안에 같이 있는 건 비즈니스 요구에 따라 필요한 절차를 묶는 것이라
책임이 섞인 게 아니라, 오히려 역할 분리를 잘 지키고 있다 라고 생각하시면 도움이 될겁니다

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -14,59 +20,66 @@
-		Item item = new Item(req.name(), req.price(), req.imageUrl());
-		db.put(item.getId(), item);
-		return item.getId();
+		final String sql = "INSERT INTO item (name, price, image_url) VALUES (?, ?, ?)";
```
상품 생성도 온전한 Item 을 만들고 그 Item 을 저장만하는 로직을 구현하는게 좋을것 같아요

silverttthin replied:
현재 로직은 쿼리에 들어가는 값을 reqDto에서 직접 가져오고 있습니다.
피드백해주신 부분들은 Item 엔티티 객체를 req를 통해 조립하고 이 객체값을 쿼리에 활용하라는 것으로 확인되는 것 같아 그렇게 구현했습니다. 
그러나 지금 로직으로도 구현과 작동, 테스트는 정상처리됩니다. jpa는 객체를 만들고 jpa 메서드에 인자로 넣긴하지만 jdbctemplate에서도 이렇게 해야하는 이유가 있을까요? 

그나마 제가 이거일것 같은 추측은.. 
reqDto는 요청 데이터를 담는 책임이고 데이터베이스에 담는 책임을 가진 객체는 엔티티객체를 따로 만들어야 한다는 것..? 이번에 해주신 피드백의 의미는 잘 모르겠습니다 😢

mskangg replied:
Service 와 Repository 의 차이점을 숙지하시고 구현해보시면 어느정도 해결될 문제 같아요!
추가적인 고민사항이 있으시면 코멘트 부탁드립니당 :)

silverttthin replied:
피드백 반영해서 다음처럼 구현했습니다!
1. dto 단에서 요청값 검증로직 돌아감
2. 서비스 단에서 req dto값을 기반으로 new Item()을 조립
3. item 값을 itemRepositoryImpl에 넘겨서 db 로직에서 활용

서비스는 비즈니스로직(유효성, 상태체크), 리포지토리는 데이터 접근의 책임을 갖는 차이가 있습니다.

구현해보며 느낀 점으론.. req.name, req.price.. 을 쿼리에 직접 활용하기 보단 item 객체를 만들어서 넘기는게 뭐랄까 맞고 이게 정론인 것 같은 느낌인데 왜 이게 맞는지 구술해보라고 하면 곧바로 하기는 어려운 것 같습니다

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -14,59 +20,66 @@
+
+		final String sql = "select * from item where id = ?";
+
+		return jdbcTemplate.queryForObject(sql, (rs, rowNum) -> new GetItemResponse(
```
Item 을 리턴해야하지 않을까요?

silverttthin replied:
> 상품 생성도 온전한 Item 을 만들고 그 Item 을 저장만하는 로직을 구현하는게 좋을것 같아요


이것도 위의 코멘트의 것과 비슷한 질문을 드리고싶습니다!
GetItemResponse라는 응답 dto가 리턴되는게 더 맞지 않나요?

mskangg replied:
https://github.com/next-step/spring-gift-product/pull/222#discussion_r2176332401

mskangg commented on README.md:
```diff
@@ -24,4 +24,14 @@
-- [X] 상품 삭제 기능 연동
\ No newline at end of file
+- [X] 상품 삭제 기능 연동
+
```
> 이번 여수 카카오 커넥트 데이에서 카카오 개발자분께 제가 다음 질문을 드리고 답변받았습니다.

코멘트 정말 공감돼요 ㅎㅎ  
저도 신입 채용할 때는 `뭘 했냐`보다 `왜 이걸 썼고, 어떻게 문제를 풀었는지` 고민한 과정, 그리고 `팀 내에서 소통하고 성장하려는 태도`를 훨씬 더 중요하게 봐요 👍  
오히려 이유 없는 신기술 도입이나 그걸 활용한 포트폴리오는 매력도가 좀 떨어지더라고요 😅

요즘엔 LLM, AI가 워낙 대중화돼서, 단순히 레이어드 아키텍처로 API 서버를 만들었다고 어필하는 건 확실히 임팩트가 약한 것 같아요 🤔  
하지만 JPA, MSA, 이벤트, 비동기 등 기술을 마구잡이로 적용한 것보단, 레이어드 아키텍처의 구조를 왜 적용했고 단점은 무엇인지 설명하고 어떤 차별점이 있으며 이 프로젝트엔 어떤 이점이 있어 선택했는지에 대한 이유가 충분하다면 그 포트폴리오의 서류를 합격시키곤 합니다 😄   
추가적으로 프로젝트 운영 규모가 커지면 어떻게 구조가 바뀔지까지 고민하고 설명까지 되어있다면 더 눈에 들어옵니다 👍 

결국 `내가 무슨 문제를 발견했고`, `왜, 어떻게 해결했는지`,  `팀원과 어떤 논의를 했는지` 이런 스토리가 훨씬 어필돼요!  
수직, 수평 확장도 물론 중요하지만,  `문제 정의 → 해결 과정 → 결과` 이 흐름을 잘 보여주는 게 신입 포트폴리오에서 제일 중요한 것 같아요 👍 

궁금한 점 있으면 언제든 편하게 질문 주세요!  
(이런 고민 자체가 이미 좋은 개발자의 시작이라고 생각합니다 😊)

mskangg reviewed (changes requested):
3단계 잘 구현해주셨네요!
질문에 대한 답변과 추가적인 코멘트 남겨드렸어요!
확인 후 재요청 부탁드립니다 🙇

> 여담이지만, 기술에 너무 매몰되지 않았으면 해요 🤔
기술을 아는 것도 중요하지만, 왜 이 기술이 나왔고, 어디에 써야 효과적인지 고민 없이 적용하는 건 의미가 없다고 생각해요 🙃
항상 기술은 `필요`에 의해 쓰여야 하고, 기술을 쓰기 위해 억지로 프로젝트에 끼워맞추는 건 지양했으면 해서 추가로 코멘트 남깁니다 😅

silverttthin committed:
feat: Item 객체를 사용하는 것으로 변경
- 기존 서비스 레이어에서 쿼리 조립시 reqDto를 직접 사용
- Item 객체를 req으로부터 조립해 만들기로
- 이를 위해 Item 자체에도 pk 유무에 따른 생성자를 하나 더 만들었음
(추후 레포지토리에 Item을 넣을 예정)

silverttthin committed:
feat: 레포지토리 인터페이스 작성

silverttthin committed:
feat: 레포지토리 구현체 작성 및 서비스 레이어와의 분리
- 단건조회를 Optional로 해 NPE 대처 및 존재확인 로직으로 사용

silverttthin commented:
위에서 해주신 피드백들을 반영하여 리포지토리 계층을 도입했습니다.
1. 처음엔 그냥 아무생각없이..? 리포지토리 클래스를 만들어 거기에서 db 작성 로직을 작성했습니다.
2. 다 완성하고 다른사람들 어떻게 했을까 pr들 구경하다가 강원대 최현준님의 pr을 보게됐습니다.
3. db 동작을 interface로 작성해둔 리포지토리와 구현체를 작성하고 service에선 구현체 빈을 가져와 사용하고 있었습니다.
4. 서비스가 구현체가 아닌 추상화에 의존하는 매우 좋은 패턴이고(서비스는 구현체보다 상위모듈이니) OOP의 다형성을 잘 활용하는 좋은 코드형식인 것 같다고 생각이 들어 이렇게 구현했습니다.

사실 아! 이래서 인터페이스를 무조건 도입해야하는구나 라고 생각이 들 정도로 효용을 크게 체감한 경험은 없어서 위의 지식들이 온전히 와닿진 않지만 좋은 패턴이고 당장은 몰라도 이렇게 알아두고 구현해두면 언젠가 체화될것 같단 생각이 들어 구현했습니다

silverttthin committed:
feat: 서비스 메서드에 트랜잭션 설정

silverttthin committed:
test: 테스트 코드 작성

mskangg commented on src/test/java/gift/ItemRepositoryTest.java:
```diff
@@ -0,0 +1,66 @@
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+
+@JdbcTest // jdbc 관련 빈만 띄울때 사용 for 속도
```
jdbcText 를 활용해주셨군요 👍

mskangg commented on src/test/java/gift/ItemRepositoryTest.java:
```diff
@@ -0,0 +1,66 @@
+@JdbcTest // jdbc 관련 빈만 띄울때 사용 for 속도
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@Import(ItemRepositoryImpl.class)
+@ActiveProfiles("test")
```
test 환경설정 👍

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -1,72 +1,85 @@
-			db.put(item.getId(), item);
-		}
-	}
+	private final ItemRepositoryImpl itemRepository;
```
> 서비스가 구현체가 아닌 추상화에 의존하는 매우 좋은 패턴이고(서비스는 구현체보다 상위모듈이니) OOP의 다형성을 잘 활용하는 좋은 코드형식인 것 같다고 생각이 들어 이렇게 구현했습니다.

인터페이스로 추상화를 잘해주셔서 DIP 를 지켜주셨는데, 직접 구현체를 의존해주셨네요 😅
여기서 인터페이스를 의존해도 잘 동작하는데, 스프링은 어떻게 구현했을까요? 🤔

silverttthin replied:
스프링은 앱 실행시 컴포넌트 스캔을 통해 우선 의존성이 없는 빈부터 먼저 생성합니다.
그리고 서비스 빈이나 컨트롤러 빈처럼 의존성이 필요한 빈의 경우 그 빈을 찾습니다. 
서비스의 경우 인터페이스에 의존하므로 인터페이스 빈을 구현하는 클래스들을 찾아와 서비스에 구현체를 때려박습니다.
(별 생각없이 구현체를 직접 의존하게 했네요 ㅠ 인터페이스에 의존해도 되는 부분을 이제 알았습니다;)

만약 구현체가 여러개일 경우 primary 어노테이션 등으로 구현체에 스프링에서 우선순위를 인식하게 하는 방법도 있다는 것을 알았습니다

mskangg commented on src/main/java/gift/product/ItemService.java:
```diff
@@ -1,72 +1,85 @@
-		item.setImageUrl(req.imageUrl());
-		return new GetItemResponse(item.getId(), item.getName(), item.getPrice(), item.getImageUrl());
+
+		itemRepository.findById(itemId)
+			.orElseThrow(() -> new RuntimeException("존재하지 않는 아이템입니다."));
```
👍

mskangg commented on src/main/java/gift/product/ItemRepositoryImpl.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.Optional;
+
+
+@Repository
+public class ItemRepositoryImpl implements ItemRepository {
```
> 사실 아! 이래서 인터페이스를 무조건 도입해야하는구나 라고 생각이 들 정도로 효용을 크게 체감한 경험은 없어서 위의 지식들이 온전히 와닿진 않지만 좋은 패턴이고 당장은 몰라도 이렇게 알아두고 구현해두면 언젠가 체화될것 같단 생각이 들어 구현했습니다

이 부분은 앞으로 더 공부하고 실무 경험이 쌓이다 보면 필요한 순간이 분명히 올 거예요.
처음에는 무작정 적용하면 오히려 불필요하게 느껴질 때도 있는데, 그럴 때마다 왜 필요한지 고민해보고, 직접 의미 있는 결론을 내리면서 점점 성장하실겁니다 😄
다만, 무분별한 추상화는 오히려 복잡도를 증대시키니 이 부분을 항상 인지하시는것도 중요해요 👍 
너무 조급하게 생각하지 않으셔도 괜찮습니다 ㅎㅎ

silverttthin replied:
감사합니다! 1단계 미션 리뷰해주시고 피드백해주신 부분들 잘 소화하고 제것으로 가져가도록 노력하겠습니다

mskangg reviewed (approved):
피드백 반영 잘해주셨네요!
고민에 대한 흔적이 느껴지니 저도 좋습니다 👍 
추가로 수정할 부분이 있어 코멘트 남겼는데, 이 부분은 답글 코멘트 정도로만 부탁드립니다 🙇
 
첫미션 진행하시느라 정말 고생 많으셨습니다.
시작이 반이라는 말도 있는것처럼 언제나 처음이 가장 힘든데, 그걸 해내셨군요 👍 
초심 잃지마시고 시작하기전 마음가짐 그대로 끝까지 미션완주하셔서 원하시는바 이루시길 바랄게요!
그간 고생 많으셨습니다 👍
